[SILGen] Fix the type of closure thunks that are passed const reference structs

ahatanaka opened the pull request:
The thunk's parameter needs the @in_guaranteed convention if it's a const reference parameter. However, that convention wasn't being used because clang importer was removing the const reference from the type and SILGen was computing the type of the parameter based on the type without const reference.

This commit fixes the bug by passing the clang function type to SILDeclRef so that it can be used to compute the correct thunk type.

This fixes a crash when a closure is passed to a C function taking a pointer to a function that has a const reference struct parameter.

rdar://131321096

<!--
If this pull request is targeting a release branch, please fill out the
following form:
https://github.com/swiftlang/.github/blob/main/PULL_REQUEST_TEMPLATE/release.md?plain=1

Otherwise, replace this comment with a description of your changes and
rationale. Provide links to external references/discussions if appropriate.
If this pull request resolves any GitHub issues, link them like so:

  Resolves <link to issue>, resolves <link to another issue>.

For more information about linking a pull request to an issue, see:
https://docs.github.com/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue
-->

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/main/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

ahatanaka committed:
[SILGen] Fix the type of closure thunks that are passed const reference
structs

The thunk's parameter needs the @in_guaranteed convention if it's a
const reference parameter. However, that convention wasn't being used
because clang importer was removing the const reference from the
type and SILGen was computing the type of the parameter based on the
type without const reference.

This commit fixes the bug by passing the clang function type to
SILDeclRef so that it can be used to compute the correct thunk type.

This fixes a crash when a closure is passed to a C function taking a
pointer to a function that has a const reference struct parameter.

rdar://131321096

ahatanaka commented:
@swift-ci please test

rjmccall commented on lib/SILGen/SILGenExpr.cpp:
```diff
@@ -1797,7 +1807,9 @@ ManagedValue emitCFunctionPointer(SILGenFunction &SGF,
   }
-  
+
+  const clang::Type *destFnType = nullptr;
```
This ought to be named something that makes it clear that this is a Clang type.

rjmccall commented on lib/SILGen/SILGenExpr.cpp:
```diff
@@ -1726,7 +1726,17 @@ static ManagedValue convertCFunctionSignature(SILGenFunction &SGF,
+  else
+    loweredDestTy = SGF.getLoweredType(
+        AbstractionPattern(destTy->getCanonicalType(), clangInfo.getType()),
+        destTy);
```
This should not be necessary.  If we're lowering a foreign function type, we should be honoring the Clang type stored in it automatically without requiring that we pass down separately as an abstraction pattern.

ahatanaka replied:
I'm passing the type in the abstraction pattern because the stored clang type gets dropped in `TypeConverter::getTypeLowering` if I just pass the type without the abstraction pattern to `getLoweredType`.

```
  const TypeLowering &
  getTypeLowering(Type t, TypeExpansionContext forExpansion) {
    AbstractionPattern pattern(t->getCanonicalType());
    return getTypeLowering(pattern, t, forExpansion);
  }
```

The call to `t->getCanonicalType()` drops the clang type.

ahatanaka replied:
Actually, `t` still has the clang type, but it's dropped somewhere else later.

rjmccall replied:
It'd be good to understand that.  I don't think there's any good reason for us to drop clang types, especially during type lowering.

ahatanaka replied:
Setting `LangOptions::UseClangFunctionTypes` or passing `-use-clang-function-types` prevents clang types from getting dropped, but it also causes other regression tests to crash/fail.

It looks like it hasn't been turned on by default because of unresolved issues.

```
    /// Use Clang function types for computing canonical types.
    /// If this option is false, the clang function types will still be computed
    /// but will not be used for checking type equality.
    /// [TODO: Clang-type-plumbing] Turn on for feature rollout.
    bool UseClangFunctionTypes = false;
```

rjmccall replied:
Well, that's a shame.  Still, are we sure the information isn't available from the decl that we already have in the SILDeclRef?

ahatanaka replied:
Do you mean the `SILDeclRef` that's declared later in the function?

```
// Produce a reference to the C-compatible entry point for the function.
SILDeclRef constant(loc, /*foreign*/ true);
```

`loc` passed to the constructor is an `AbstractClosureExpr`, whose type doesn't have a clang type. The type of `FunctionConversionExpr` passed to this function has the clang type.

rjmccall replied:
I still think this shouldn't be necessary — lowering a `@convention(c)` function type that has a Clang type attached using it as its own abstraction pattern should definitely be producing a type that uses the conventions of that Clang type.  Can you identify what exactly drops this?  It's one thing if it's just not being propagated into the `SILFunctionType`, but it seems bizarre that we'd not use this information when it's available in type lowering.

ahatanaka replied:
The clang type is dropped in `TypeConverter::getTypeLowering`.

```
  getTypeLowering(Type t, TypeExpansionContext forExpansion) {
    AbstractionPattern pattern(t->getCanonicalType());
    return getTypeLowering(pattern, t, forExpansion);
  }
```

The call to `t->getCanonicalType()` drops the clang type because `UseClangFunctionTypes` isn't set.

I experimented with creating an `AbstractionPattern` with a clang type in that function if it exists and it seemed to work although I didn't test it extensively.

ahatanaka replied:
And the type `t` passed to the overload of `getTypeLowering` loses its clang type when `getCanonicalType` is called again.

```
const TypeLowering &
TypeConverter::getTypeLowering(AbstractionPattern origType,
                               Type origSubstType,
                               TypeExpansionContext forExpansion) {
  CanType substType = origSubstType->getCanonicalType();
```

rjmccall replied:
Ah, okay.  You'd said up-thread that it wasn't dropped by canonicalization, but if it is, I agree that we can't do much here.  Please at least add a comment explaining that this won't be necessary when we switch to Clang types, because the Clang function type should survive canonicalization.

rjmccall reviewed (commented):
The basic approach here seems wrong to me.  When lowering a foreign function type, or the type of a foreign SILDeclRef, we should be able to recover its foreign type without expecting it to be passed down separately.

ahatanaka committed:
Add a precondition to SILDeclRef's constructor

The precondition checks that a clang type must be passed when creating
the foreign entry point for a closure.

ahatanaka committed:
Add test cases for testing closure-to-block conversion

ahatanaka commented:
@swift-ci please test

ahatanaka committed:
Add a comment

rjmccall reviewed (approved):
Alright, thanks, LGTM.

ahatanaka commented:
@swift-ci please smoke test

ahatanaka committed:
Fix size of bridged classes

This is needed to fix the static assertions in the file that were
failing because a new member was added to SILDeclRef.

ahatanaka commented:
@swift-ci please test

ahatanaka commented:
@swift-ci Please Build Toolchain macOS Platform
